### PR TITLE
Identity 3132 - 'Key length exceeded' error on CON_APP_KEY of IDN_OAUTH2_ACCESS_TOKEN table in ORACLE DB

### PR DESCRIPTION
**Pull request by madurangasiriwardena**

### CODE REVIEW

File: components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java [hunk 1]
```diff
@@ -148,6 +148,15 @@ private TokenStates(){
         }
     }
 
+    public static class AuthorizationCodeState {
+        public static final String ACTIVE = "ACTIVE";
+        public static final String EXPIRED = "EXPIRED";
```
Comment (hpmtissera) [inline components/oauth/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/OAuthConstants.java]: add new state Inactive

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AuthPersistenceTask.java [hunk 2]
```diff
@@ -48,7 +48,8 @@ public void run() {
                     if (authContextTokenDO.getAuthzCodeDO() == null) {
                         log.debug("Auth Token Data removing Task is started to run");
                         TokenMgtDAO tokenMgtDAO = new TokenMgtDAO();
-                        tokenMgtDAO.removeAuthzCode(authContextTokenDO.getAuthzCode());
+                        tokenMgtDAO.doExpireAuthorizationCode(authContextTokenDO.getAuthzCode(), authContextTokenDO
```
Comment (hpmtissera) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AuthPersistenceTask.java]: deactivate

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 3]
```diff
@@ -800,6 +879,45 @@ public void revokeToken(String token) throws IdentityOAuth2Exception {
...
     }
 
+
+    /**
+     * This method is to revoke specific tokens
+     *
+     * @param tokenId token that needs to be revoked
+     * @throws IdentityOAuth2Exception if failed to revoke the access token
+     */
+    public void revokeToken(String tokenId, String userId) throws IdentityOAuth2Exception {
```
Comment (hpmtissera) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: add token id to dao

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 4]
```diff
@@ -176,15 +181,20 @@ public void storeAccessToken(String accessToken, String consumerKey,
...
+
         String sql = "INSERT INTO " +
                      accessTokenStoreTable +
-                     " (ACCESS_TOKEN, REFRESH_TOKEN, CONSUMER_KEY, AUTHZ_USER, TIME_CREATED, REFRESH_TOKEN_TIME_CREATED, " +
-                     "VALIDITY_PERIOD, REFRESH_TOKEN_VALIDITY_PERIOD, TOKEN_SCOPE, TOKEN_STATE, USER_TYPE) " +
-                     "VALUES (?,?,?,?,?,?,?,?,?,?,?)";
+                " (ACCESS_TOKEN, REFRESH_TOKEN, CONSUMER_KEY, AUTHZ_USER, TIME_CREATED, REFRESH_TOKEN_TIME_CREATED, " +
+                "VALIDITY_PERIOD, REFRESH_TOKEN_VALIDITY_PERIOD, TOKEN_SCOPE_HASH, TOKEN_STATE, USER_TYPE, TOKEN_ID) " +
+                "VALUES (?,?,?,?,?,?,?,?,?,?,?,?)";
+        String sqlAddScopes = "INSERT INTO IDN_OAUTH2_ACCESS_TOKEN_SCOPE_ASSOCIATION (TOKEN_ID, TOKEN_SCOPE) " +
```
Comment (hpmtissera) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: move this to SQLQueries

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 5]
```diff
@@ -196,13 +206,23 @@ public void storeAccessToken(String accessToken, String consumerKey,
...
             prepStmt.setLong(7, accessTokenDO.getValidityPeriodInMillis());
             prepStmt.setLong(8, accessTokenDO.getRefreshTokenValidityPeriodInMillis());
-            prepStmt.setString(9, OAuth2Util.buildScopeString(accessTokenDO.getScope()));
+            prepStmt.setString(9, OAuth2Util.hashScopes(accessTokenDO.getScope()));
             prepStmt.setString(10, accessTokenDO.getTokenState());
             prepStmt.setString(11, accessTokenDO.getTokenType());
+            prepStmt.setString(12, accessTokenDO.getTokenId());
             prepStmt.execute();
+            String accessTokenId = accessTokenDO.getTokenId();
+            prepStmt = connection.prepareStatement(sqlAddScopes);
```
Comment (hpmtissera) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: check this

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 6]
```diff
@@ -200,14 +208,24 @@ public void storeAccessToken(String accessToken, String consumerKey,
...
             prepStmt.setLong(8, accessTokenDO.getRefreshTokenValidityPeriodInMillis());
-            prepStmt.setString(9, OAuth2Util.buildScopeString(accessTokenDO.getScope()));
+            prepStmt.setString(9, OAuth2Util.hashScopes(accessTokenDO.getScope()));
             prepStmt.setString(10, accessTokenDO.getTokenState());
             prepStmt.setString(11, accessTokenDO.getTokenType());
             prepStmt.setString(12, accessTokenDO.getTokenId());
             prepStmt.execute();
+            String accessTokenId = accessTokenDO.getTokenId();
+            prepStmt = connection.prepareStatement(sqlAddScopes);
+            for (String scope : accessTokenDO.getScope()){
```
Comment (omindu) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Formatting issue.

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 7]
```diff
@@ -630,16 +662,26 @@ public RefreshTokenValidationDataDO validateRefreshToken(String consumerKey, Str
...
+                    validationDataDO.setIssuedTime(resultSet.getTimestamp(5, Calendar.getInstance(TimeZone.getTimeZone
+                            ("UTC"))));
+                    validationDataDO.setValidityPeriodInMillis(resultSet.getLong(6));
+                    validationDataDO.setTokenId(resultSet.getString(7));
+                } else {
+                    scopes.add(resultSet.getString(3));
+                }
+                iterateId++;
+            }
+            if (scopes.size() > 0){
```
Comment (omindu) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Formatting issue.

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 8]
```diff
@@ -515,6 +516,15 @@ public static int getTenantIdFromUserName(String username) throws IdentityOAuth2
         return getTenantId(domainName);
     }
 
+    public static String hashScopes(String[] scope){
+        return DigestUtils.md5Hex(OAuth2Util.buildScopeString(scope));
+    }
+
+    public static String hashScopes(String scope){
```
Comment (omindu) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Formatting issue.

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 9]
```diff
@@ -515,6 +516,15 @@ public static int getTenantIdFromUserName(String username) throws IdentityOAuth2
         return getTenantId(domainName);
     }
 
+    public static String hashScopes(String[] scope){
```
Comment (omindu) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Formatting issue.

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java [hunk 10]
```diff
@@ -731,24 +773,34 @@ public AccessTokenDO retrieveAccessToken(String accessTokenIdentifier, boolean i
...
+                            validityPeriodInMillis, refreshTokenValidityPeriodMillis, tokenType);
+                    dataDO.setAccessToken(accessTokenIdentifier);
+                    dataDO.setRefreshToken(refreshToken);
+                    dataDO.setTokenId(tokenId);
+                }else{
+                    scopes.add(resultSet.getString(3));
+                }
+                iterateId++;
+            }
+            if (scopes.size() > 0){
```
Comment (omindu) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Formatting issue.

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java (hunk 6 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java (hunk 7 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java (hunk 8 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java (hunk 9 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

File: components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java (hunk 10 above)
Comment (madurangasiriwardena) [inline components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1679

**Pull request merged**